### DISCUSSION TITLE
Reading L2 state from L1.

### DISCUSSION THREAD

**Discussion by Bapi-Reddy** [ethereum-optimism/optimistic-specs#363] [Q&A]
Is there a way to verify L2 contract state at specific block on L1 using merkle proofs of some form. Can the state tries from `State Commitment Chain` be helpful in some form. The idea is if there is a decentralized way to verify state of L2 contracts on L1 it increases the support cross chain oracles which can be used for enabling crossh-chain defi apps.

**Comment by norswap** [comment]
It's easy to prove against the latest L1 block that L2 build upon via the L1 attributes contract:
- https://github.com/ethereum-optimism/optimistic-specs/blob/main/specs/deposits.md#l1-attributes-predeployed-contract
- https://github.com/ethereum-optimism/optimistic-specs/blob/main/packages/contracts/contracts/L2/L1Block.sol

It's theoretically possible this use the blockhash there to also build a proof for a specific L1 block (via the parent hash), but the proof gets too long fairly quickly.

The solution there would be store a Merkle root of a trie that stores the `block height → blockhash` mapping for L1 blocks. This is something we've discussed, and I think this is something we want to do (as you say, there are quite some use cases). Feel free to open an issue to track this :)

**Reply by Bapi-Reddy** [reply]
I think you understood my question incorrectly. I want to verify the state of L2 contract on L1. So its the other away around. Where a I want to check if merkle proof for l2 contract variable on latest committed block is valid on L1.

**Reply by trianglesphere** [reply]
You would use the L2 Output Oracle contract (this is how withdrawals are done) on L1.

**Reply by Bapi-Reddy** [reply]
I see its still in proposal stage here:[https://github.com/ethereum-optimism/optimistic-specs/blob/main/packages/contracts/contracts/L1/L2OutputOracle.sol ]( https://github.com/ethereum-optimism/optimistic-specs/blob/main/packages/contracts/contracts/L1/L2OutputOracle.sol )
or is it live now ? . Are there any code examples showing its usage ?

**Reply by trianglesphere** [reply]
This repo is not yet live. That's the correct contract accessing the state for bedrock. You can see withdrawals being verified here: https://github.com/ethereum-optimism/optimistic-specs/blob/main/packages/integration-tests/test/000_withdrawals.spec.ts
You could look at `withdrawalRelay.sol` for an example of performing the verification of some L2 state on L1. FYI these contract's are still pre-production.

It's probably a similar process for the current `State Commitment Chain`
